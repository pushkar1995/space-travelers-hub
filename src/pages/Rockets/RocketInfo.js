import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { reserveRocket, cancelRocket } from '../../redux/rocketSlice';

const RocketInfo = ({
  name, description, image, id, reserved,
}) => {
  const dispatch = useDispatch();

  const toggleReservedStatus = () => {
    if (reserved) {
      dispatch(cancelRocket(id));
    } else {
      dispatch(reserveRocket(id));
    }
  };

  const buttonClass = reserved ? 'reserved-button' : 'reserve-rocket';

  return (
    <div className="rockets-container">
      <div>
        <img className="rockets-pictures" src={image[0]} alt="rockets-pictures" />
      </div>
      <div className="rockets-info">
        <h2 className="rocket-name">{name}</h2>
        <p>
          {reserved ? <span className="reservation-status">Reserved</span> : null}
          {description}
        </p>
        <button
          type="button"
          className={buttonClass}
          onClick={toggleReservedStatus}
        >
          {reserved ? 'Cancel Reservation' : 'Reserve Rocket'}
        </button>
      </div>
    </div>
  );
};

export default RocketInfo;

RocketInfo.propTypes = {
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  image: PropTypes.arrayOf(PropTypes.string).isRequired,
  id: PropTypes.string.isRequired,
  reserved: PropTypes.bool,
};

RocketInfo.defaultProps = {
  reserved: false,
};
