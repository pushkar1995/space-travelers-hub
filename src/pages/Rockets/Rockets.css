.individual-rocket {
    padding: 2rem;
    display: flex;
  }
  
  .about-rocket {
    padding: 1rem;
  }
  
  .reserve-btn {
    padding: 0.7rem;
    background-color: #0260ff;
    color: #fff;
  }
  
  .reserve-btn:hover {
    background-color: #014ed3;
  }
  