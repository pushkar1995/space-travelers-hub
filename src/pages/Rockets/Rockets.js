import { useSelector } from 'react-redux';
import RocketInfo from './RocketInfo';
import './Rockets.css';

const Rockets = () => {
  const { rockets } = useSelector((store) => store.rockets);

  return (
    <div>
      {rockets.map((rocket) => (
        <RocketInfo
          key={rocket.id}
          id={rocket.id}
          name={rocket.name}
          description={rocket.description}
          image={rocket.flickr_images}
          reserved={rocket.reserved}
        />
      ))}
    </div>
  );
};

export default Rockets;
