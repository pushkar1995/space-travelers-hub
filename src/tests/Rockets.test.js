import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Rockets from '../pages/Rockets/Rockets';

jest.mock('../redux/rocketSlice');

const mockStore = configureStore([]);

describe('Rockets Component', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      rockets: {
        status: 'succeeded',
        error: null,
      },
    });
  });

  it(' it renders loading state', () => {
    store = mockStore({
      rockets: {
        status: 'loading',
        error: null,
      },
    });

    const { container } = render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders error state', () => {
    store = mockStore({
      rockets: {
        status: 'failed',
        error: 'Test Error',
      },
    });

    const { container } = render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });
});
