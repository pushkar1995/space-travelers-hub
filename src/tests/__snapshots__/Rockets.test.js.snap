// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets component renders correctly 1`] = `
<div>
  <div>
    <div
      class="rockets-container"
    >
      <div>
        <img
          alt="rockets-pictures"
          class="rockets-pictures"
          src="image1.jpg"
        />
      </div>
      <div
        class="rockets-info"
      >
        <h2
          class="rocket-name"
        >
          Falcon 9
        </h2>
        <p>
          A powerful rocket developed by SpaceX.
        </p>
        <button
          class="reserve-rocket"
          type="button"
        >
          Reserve Rocket
        </button>
      </div>
    </div>
  </div>
</div>
`;
